### DISCUSSION TITLE
Keycloak configuration in the CRD

### DISCUSSION THREAD

**Discussion by andreaTP** [keycloak/keycloak#11848] [Archive]
Reference issue for this discussion: https://github.com/keycloak/keycloak/issues/10866

# Goal

We want to give people a great experience when configuring Keycloak by using the [Keycloak CRD](https://github.com/keycloak/keycloak-k8s-resources/blob/main/kubernetes/keycloaks.k8s.keycloak.org-v1.yml)

## Nice to have

When adding/changing configuration in the Quarkus distribution, changes are aligned to the CRD.

### Current status

The current implementation relies on a ["free-form" map of values](https://github.com/keycloak/keycloak-k8s-resources/blob/153d9308503380fa076780a01669e4c9751a0c69/kubernetes/keycloaks.k8s.keycloak.org-v1.yml#L22-L42).

PROS:
  - great flexibility
  - reads nicely along with other Kubernetes settings
  
CONS:
  - no checks were performed when applying the CRD to Kubernetes
  - no validation

I went ahead and prepared a list of the [current configuration options and how they might look like in a CRD](https://gist.github.com/andreaTP/d2baf122fdc71793d1d66f42f5ae89c6).

### Proposals

Here I present two "extreme" alternatives, probably we can evaluate something in-between those.

#### Automatic CRD generation

In [this Draft PR](https://github.com/keycloak/keycloak/pull/11806) I have explored the possibility of automatically generating the configuration out of the `PropertyMappers` in the Quarkus distribution CLI options.
Regardless of the technical implementation of it, that remains open, I found a few challenges to achieving the desired UX:

- the configuration keys for the CLI are in Kebab Case but the rest of the CRD is using Camel Case
- there are configuration options that do not make sense to expose knowing we are on Kubernetes
- there are configuration options that might be mapped differently on Kubernetes (e.g. file names might be translated to secret/configmap references)

#### Manual CRD crafting

Manually integrating the configuration options, since they are not "uncountable", is still a viable option at this point in time.
Challenges will be especially in maintenance:

- no guarantees about the sync between CLI and CRD options
- possibly no single source of truth

##### Additional notes

Build time options vs. runtime options, ideally, build-time options should not be available in case a custom(pre-augmented) Keycloak image is provided.
Possibly a [validating admission webhook](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#validatingadmissionwebhook) can improve the user experience in this regard.

**Comment by andreaTP** [comment]
Follow up.
This is the most relevant POC regarding the subject: https://github.com/keycloak/keycloak/pull/11929

We explored the automatic generation a little more and we are starting to achieve some kind of agreement.
Those are the answer to the open questions:

> - the configuration keys for the CLI are in Kebab Case but the rest of the CRD is using Camel Case

automatically converting to Camel Case for the CRD

> - there are configuration options that do not make sense to expose knowing we are on Kubernetes

have runtime specific filters

> - there are configuration options that might be mapped differently on Kubernetes

specifically handled in the generator

**Comment by vmuzikar** [comment]
Related discussion: https://github.com/keycloak/keycloak/discussions/13682
